homeassistant/helpers/restore_state.py

```diff
@@ -1,98 +1,128 @@
-
-    last_run = last_recorder_run(hass)
-
-    if last_run is None or last_run.end is None:
-        _LOGGER.debug('Not creating cache - no suitable last run found: %s',
-                      last_run)
-        hass.data[DATA_RESTORE_CACHE] = {}
-        return
-
-    last_end_time = last_run.end - timedelta(seconds=1)
-    # Unfortunately the recorder_run model do not return offset-aware time
-    last_end_time = last_end_time.replace(tzinfo=dt_util.UTC)
-    _LOGGER.debug("Last run: %s - %s", last_run.start, last_end_time)
-
-    states = get_states(hass, last_end_time, run=last_run)
-
-    # Cache the states
-    hass.data[DATA_RESTORE_CACHE] = {
-        state.entity_id: state for state in states}
-    _LOGGER.debug('Created cache with %s', list(hass.data[DATA_RESTORE_CACHE]))
-
-
-@bind_hass
-async def async_get_last_state(hass, entity_id: str):
-    """Restore state."""
-    if DATA_RESTORE_CACHE in hass.data:
-        return hass.data[DATA_RESTORE_CACHE].get(entity_id)
-
-    if _RECORDER not in hass.config.components:
-        return None
-
-    if hass.state not in (CoreState.starting, CoreState.not_running):
-        _LOGGER.debug("Cache for %s can only be loaded during startup, not %s",
-                      entity_id, hass.state)
-        return None
-
-    try:
-        with async_timeout.timeout(RECORDER_TIMEOUT, loop=hass.loop):
-            connected = await wait_connection_ready(hass)
-    except asyncio.TimeoutError:
-        return None
-
-    if not connected:
-        return None
-
-    if _LOCK not in hass.data:
-        hass.data[_LOCK] = asyncio.Lock(loop=hass.loop)
-
-    async with hass.data[_LOCK]:
-        if DATA_RESTORE_CACHE not in hass.data:
-            await hass.async_add_job(
-                _load_restore_cache, hass)
-
-    return hass.data.get(DATA_RESTORE_CACHE, {}).get(entity_id)
-
-
-async def async_restore_state(entity, extract_info):
-    """Call entity.async_restore_state with cached info."""
-    if entity.hass.state not in (CoreState.starting, CoreState.not_running):
-        _LOGGER.debug("Not restoring state for %s: Hass is not starting: %s",
-                      entity.entity_id, entity.hass.state)
-        return
-
-    state = await async_get_last_state(entity.hass, entity.entity_id)
-
-    if not state:
-        return
-
-    await entity.async_restore_state(**extract_info(state))
+    @callback
+    def async_register_entity(
+            self: 'RestoreStateData', entity: 'RestoreEntity') -> None:
+        """Store this entities state when hass is shutdown."""
+        self.entities.append(entity)
+
+
+class RestoreEntity(Entity):
+    """Mixin class for restoring previous entity state."""
+
+    async def async_added_to_hass(self: 'RestoreEntity') -> None:
+        """Register this entity as a restorable entity."""
+        await super().async_added_to_hass()
+        data = await RestoreStateData.async_get_instance(self.hass)
+        data.async_register_entity(self)
```

Comment:
Why pass the whole object?

Reply:
Since `entity_id`s can change at runtime, it seemed like the easiest solution was to just hold a reference to all the entity objects, and just get the IDs when we go to dump states.

Reply:
interesting. but then you also need to add a remove listener to forget about the entity.

Reply:
Since both `async_added_to_hass` and `async_will_remove_from_hass` will be called when an entity_id is changed via the entity registry, we don't need to store the entity object in the restore state registry, but `entity_id` is enough. It's good to keep entity objects a platform concern.

Just make sure to remove the entity_id from the state restore registry in `async_will_remove_from_hass`.